### PR TITLE
Docs: clarify has_schema_privilege regarding unknown schemas

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Fixes the following scenario:
```
cr> select pg_catalog.has_schema_privilege('unknown_schema', 'usage');                                                                                                   
+------+
| true |
+------+
| TRUE |
+------+
=> should throw a SchemaUnknownException instead of returning true
```

update: instead of throwing, it was decided to return false to prevent information leaks.

update2: below is the final decision:
```
   For unknown schemas:

     - Returns ``TRUE`` for superusers.

     - For a user with ``DQL`` on cluster scope, returns ``TRUE`` if the
       privilege type is ``USAGE``.

     - For a user with ``DML`` on cluster scope, returns ``TRUE`` if the
       privilege type is ``CREATE``.

     - Returns ``FALSE`` otherwise.
```
Please follow the discussions in the current PR for details.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by seut** [comment]
To avoid an information leak, wouldn't it be better to return `False` instead? Maybe lets double check also with PG if there is a valid reason to throw an exception.

**Comment by matriv** [comment]
Postgres returns an error:
```
matriv=# select has_schema_privilege('public', 'CREATE');
 has_schema_privilege
----------------------
 t
(1 row)

matriv=# select has_schema_privilege('invalid', 'CREATE');
ERROR:  schema "invalid" does not exist

matriv=# select has_table_privilege('t', 'SELECT');
 has_table_privilege
---------------------
 t
(1 row)

matriv=# select has_table_privilege('invalid', 'SELECT');
ERROR:  relation "invalid" does not exist
```
But now that @seut mentions the information leak which we try to avoid in other cases, maybe we shouldn't change the behavior.

**Comment by jeeminso** [comment]
Will double check `has_table_privilege` and `has_database_privilege` to make the behaviour consistent.

**Comment by matriv** [comment]
Thx @jeeminso for working on this. I have some new considerations now:

What happens if the user is assigned privileges to a schema that doesn't exist yet? (we allow this through grants or restoring users & privileges from a snapshot). `sys.privileges` will show privileges on the schema for the user, but this function will return false because the schema doesn't exist. I think this is weird situation, I don't know if we want this.

**Comment by jeeminso** [comment]
> Thx @jeeminso for working on this. I have some new considerations now:
> 
> What happens if the user is assigned privileges to a schema that doesn't exist yet? (we allow this through grants or restoring users & privileges from a snapshot). `sys.privileges` will show privileges on the schema for the user, but this function will return false because the schema doesn't exist. I think this is weird situation, I don't know if we want this.

Thank you for looking in detail. I agree, then I think I can handle it in two scenarios, 1) unknown schema but granted(return true) 2) unknown schema and not granted(return false).

**Comment by matriv** [comment]
One more remark, for a non-superuser the function returns `false` currently for a schema that is not existing in the user privileges:

```
~/crate/crate ➭ ~/venv/bin/crash --hosts localhost:4200                                     
CONNECT OK
cr> create user matriv with password 'matriv';
CREATE OK, 1 row affected (0.118 sec)
cr>
Bye!
~/crate/crate ➭ ~/venv/bin/crash --hosts localhost:4200 -U matriv
Password:
CONNECT OK
cr> select has_schema_privilege('invalid', 'CREATE');
+-------+
| false |
+-------+
| FALSE |
+-------+
SELECT 1 row in set (2.293 sec)
```

It returns true only for `crate` superuser because even if it he's not assigned explicit privileges on the uknown schema, because he is superuser the `hasPrivilege(....)` call returns true:
```
~/crate/crate ➭ ~/venv/bin/crash --hosts localhost:4200                                                                               (mt/number_routing_shards|✔)
CONNECT OK
cr> select has_schema_privilege('invalid', 'CREATE');
+------+
| true |
+------+
| TRUE |
+------+
SELECT 1 row in set (2.565 sec)
```

I'd suggest to rediscuss this before changing the behavior.

**Comment by matriv** [comment]
> Thank you for looking in detail. I agree, then I think I can handle it in two scenarios, 1) unknown schema but granted(return true) 2) unknown schema and not granted(return false).

This sounds good, but we need to mention it in docs.

What do we do for superuser though, who possibly doesn't have explicit grants?
Maybe through an exception for non-existing schema in that case? (It's not leaking info, since it's the super user).

**Comment by jeeminso** [comment]
> What do we do for superuser though, who possibly doesn't have explicit grants?

Consider a case where `user_a` is granted `dql` on `schema_x` which does not exist. If `has_schema_privilege('user_a', 'schema_x', 'usage')` returns `true` maybe `has_schema_privilege('crate', 'schema_x', 'usage')` should also return `true`. Equivalently, another user with `dql` on `cluster` for example should also return `true`.

**Comment by jeeminso** [comment]
> 1) unknown schema but granted(return true) 2) unknown schema and not granted(return false)

This actually can be summarized into granted/not granted regardless of known/unknown schema -> do not need Schemas -> need to revert https://github.com/crate/crate/pull/16044

**Comment by matriv** [comment]
- I tend to keep the behavior as is for normal users, and refine the docs.
- Regarding superuser, I believe we should throw exception, as there is no info-leak to consider, and returning true for any schema, seems weird.

**Comment by jeeminso** [comment]
I agree, then the direction sounds more like an improvement than a bugfix. Should we revert https://github.com/crate/crate/pull/16044 and treat this as an improvement?

**Comment by matriv** [comment]
> I agree, then the direction sounds more like an improvement than a bugfix. Should we revert #16044 and treat this as an improvement?

Imho, sounds good, (improvement and not a bugfix).

**Comment by jeeminso** [comment]
Just to double check, I would follow the logic, _give out as much info as possible without info-leak_. For example, `has_schema_privilege('unknown_schema', 'usage')`, `dql on cluster` would see an exception but `ddl on cluster` would see `false`.

**Comment by matriv** [comment]
> Just to double check, I would follow the logic, _give out as much info as possible without info-leak_. For example, `has_schema_privilege('unknown_schema', 'usage')`, `dql on cluster` would see an exception but `ddl on cluster` would see `false`.

I would say yes, @seut do you agree?

**Comment by seut** [comment]
> Just to double check, I would follow the logic, give out as much info as possible without info-leak. For example, has_schema_privilege('unknown_schema', 'usage'), dql on cluster would see an exception but ddl on cluster would see false.

Actually, as we do not support explicit schema creation, I'm not sure if throwing an unknown schema exception for the super user is really worth it. I'd currently tend to leave it as it is and maybe revise once we support explicit schema creation...

**Comment by matriv** [comment]
> Actually, as we do not support explicit schema creation, I'm not sure if throwing an unknown schema exception for the super user is really worth it. I'd currently tend to leave it as it is and maybe revise once we support explicit schema creation...

Still, I believe it's a bit weird to return true, for the superuser, for any schema (existing, non-existing), but I'm fine to leave it as is, but @jeeminso please make sure it's clear in the docs, thx a lot!

File: docs/general/builtins/scalar-functions.rst [hunk 1]
```diff
@@ -3962,6 +3962,17 @@ Example::
...
+
+    For unknown schemas:
+
+    - Returns ``TRUE`` for superusers.
+
+    - For a user with ``DQL`` on cluster scope, returns ``TRUE`` if the
+      privilege type is ``USAGE``.
+
+    - For a user with ``DML`` on cluster scope, returns ``TRUE`` if the
+      privilege type is ``CREATE``.
```
Comment (matriv) [inline docs/general/builtins/scalar-functions.rst:3975]: I would add one more entry, to say that it returns FALSE in other cases, just to make it 100% clear.

**Review by matriv — COMMENTED** [review]
Thx a lot @jeeminso for your patience during discussions and the back and forth decisions.

LGTM. Left a comment, and please also update the description of the PR to show what we finally decided to do, thx!

File: docs/general/builtins/scalar-functions.rst (hunk 1 above)
Comment (jeeminso) [reply docs/general/builtins/scalar-functions.rst:3975]: Thank you added

**Comment by jeeminso** [comment]
> Actually, as we do not support explicit schema creation, I'm not sure if throwing an unknown schema exception for the super user is really worth it. I'd currently tend to leave it as it is and maybe revise once we support explicit schema creation...

Since we explicitly create tables, should we throw for `has_table_privilege`? (I have a [draft PR](https://github.com/crate/crate/pull/16068/commits/ba96d76b220f1683b22f7a1d8695d26cfb15f941#diff-3bc14a0e641a4cf98d82e13f93274b30b0aecc401e8936942a57273d36599994R127) with tests documenting the current behaviour)

**Review by matriv — APPROVED** [review]
thank you!

**Comment by matriv** [comment]
> > Actually, as we do not support explicit schema creation, I'm not sure if throwing an unknown schema exception for the super user is really worth it. I'd currently tend to leave it as it is and maybe revise once we support explicit schema creation...
> 
> Since we explicitly create tables, should we throw for `has_table_privilege`? (I have a [draft PR](https://github.com/crate/crate/pull/16068/commits/ba96d76b220f1683b22f7a1d8695d26cfb15f941#diff-3bc14a0e641a4cf98d82e13f93274b30b0aecc401e8936942a57273d36599994R127) with tests documenting the current behaviour)

I'd go for consistency in behaviour, so if table doesn't exist, or you don't have privilege for the passed arg, we return false. If you have privileges you get true.  For superuser you get true always (no matter if the table exists or not?) Still I find it a bit weird, @seut what do you say?

**Comment by seut** [comment]
> I'd go for consistency in behaviour, so if table doesn't exist, or you don't have privilege for the passed arg, we return false. If you have privileges you get true. For superuser you get true always (no matter if the table exists or not?) Still I find it a bit weird, @seut what do you say?

I agree to be consistent, this will avoid any information leak about non-existent tables.

**Comment by jeeminso** [comment]
Thank you @matriv @seut

**Pull request merged**